File: Lab5/Lab5dotnet/TDate.cs
using System;
using System.Linq;

namespace Lab5
{
    public abstract class TDate
    {
        protected int day;
        protected int month;
        protected int year;

        public int Day
        {
            set => day = value;
            get => day;
        }
        public int Month
        {
            set => month = value;
            get => month;
        }
        public int Year
        {
            set => year = value;
            get => year;
        }

        public int CompareTo(string date)
        {
            string[] numbers;
            int day1 = 0;
            int month1 = 0;
            int year1 = 0;
            if (date.Contains('.'))
            {
                numbers = date.Split('.');
                day1 = int.Parse(numbers[0]);
                month1 = int.Parse(numbers[1]);
                year1 = int.Parse(numbers[2]);

            }else if (date.Contains('-'))
            {
                numbers = date.Split('-');
                day1 = int.Parse(numbers[1]);
                month1 = int.Parse(numbers[0]);
                year1 = int.Parse(numbers[2]);
            }

            if (this.year < year1 ||
                this.year == year1 && (this.month < month1 || this.month == month1 && this.day < day1)) return -1;
            if (this.year == year1 && this.month == month1 && this.day == day1) return 0;
            return 1;
        }

        public void AddYears(int count)
        {
            if (count < 0)
            {
                throw new Exception("unvalid value");
            }
            this.year += count;
        }

        public void AddMonth(int count)
        {
            if (count < 0)
            {
                throw new Exception("unvalid value");
            }
            int years = count / 12;
            int remainder = count % 12;
            this.year += years;
            this.month += remainder;
            if (this.month > 12)
            {
                this.year++;
                this.month -= 12;
            }
        }

        public void AddDays(int count)
        {
            if (count < 0)
            {
                throw new Exception("unvalid value");
            }
            int years = count / 365;
            int month = (count % 365) / 30;
            int days = (count % 365) % 30;
            this.year += years;
            this.month += month;
            if (this.month > 12)
            {
                this.year++;
                this.month -= 12;
            }

            this.day += days;
            if (this.day > 30)
            {
                if (this.month < 12)
                {
                    this.month++;
                }
                else
                {
                    this.year++;
                }

                this.day -= 30;
            }
        }

        public void SubstractYears(int count)
        {
            if (count < 0)
            {
                throw new Exception("unvalid value");
            }

            this.year -= count;
        }

        public void SubstractMonth(int count)
        {
            if (count < 0)
            {
                throw new Exception("unvalid value");
            }

            int years = count / 12;
            int month = count % 12;
            this.year -= years;
            this.month -= month;
            if (month < 0)
            {
                this.year--;
                this.month += 12;
            }
        }

        public void SubstractDays(int count)
        {
            if (count < 0)
            {
                throw new Exception("unvalid value");
            }
            int years = count / 365;
            int month = (count % 365) / 30;
            int days = (count % 365) % 30;
            this.year -= years;
            this.month -= month;
            if (this.month < 0)
            {
                this.year--;
                this.month += 12;
            }

            this.day -= days;
            if (this.day < 30)
            {
                if (this.month == 1)
                {
                    this.month = 12;
                    this.year--;
                }

                this.day += 30;
            }
            
        }

        
    }

    public class TDate1 : TDate
    {
        public TDate1(string date)
        {
            string[] numbers = date.Split('.');
            this.day = int.Parse(numbers[0]);
            this.month = int.Parse(numbers[1]);
            this.year = int.Parse(numbers[2]);
        }

        public override string ToString()
        {
            return $"{this.day}.{this.month}.{this.year}";
        }
    }

    public class TDate2 : TDate
    {
        public TDate2(string date)
        {
            string[] numbers = date.Split('-');
            this.day = int.Parse(numbers[1]);
            this.month = int.Parse(numbers[0]);
            this.year = int.Parse(numbers[2]);
        }
        public override string ToString()
        {
            return $"{this.month}-{this.day}-{this.year}";
        }
    }
    
}

File: Lab5/Lab5py/TDate.py
class TDate:
    def __init__(self):
        self._day = 0
        self._month = 0
        self._year = 0

    def compare_to(self, date):
        day1 = 0
        month1 = 0
        year1 = 0
        if '.' in date:
            numbers = date.split('.')
            day1 = int(numbers[0])
            month1 = int(numbers[1])
            year1 = int(numbers[2])
        elif '-' in date:
            numbers = date.split('-')
            day1 = int(numbers[1])
            month1 = int(numbers[0])
            year1 = int(numbers[2])
        if self._year < year1 or (
                self._year == year1 and (self._month < month1 or self._month == month1 and (self._day < day1))):
            return -1
        if self._year == year1 and self._month == month1 and self._day == day1:
            return 0
        return 1

    def add_years(self, count):
        if count < 0:
            raise Exception("unvalid value")
        self._year += count

    def add_month(self, count):
        if count < 0:
            raise Exception("unvalid value")
        years = count // 12
        remainder = count % 12
        self._year += years
        self._month += remainder
        if self._month > 12:
            self._year += 1
            self._month -= 12

    def add_days(self, count):
        if count < 0:
            raise Exception("unvalid value")
        years = count // 365
        month_ = count % 365 // 30
        days = (count % 365) % 30
        self._year += years
        self._month += month_
        if self._month > 12:
            self._year += 1
            self._month -= 12
        self._day += days
        if self._day > 30:
            if self._month < 12:
                self._month += 1
            else:
                self._year += 1
            self._day -= 30

    def substract_years(self, count):
        if count < 0:
            raise Exception("unvalid value")
        self._year -= count

    def substract_month(self, count):
        if count < 0:
            raise Exception("unvalid value")
        years = count // 12
        month_ = count % 12
        self._year -= years
        self._month -= month_
        if month_ < 0:
            self._year -= 1
            self._month += 12

    def substract_days(self, count):
        if count < 0:
            raise Exception("unvalid value")
        years = count // 365
        month_ = (count % 365) // 30
        days = (count % 365) % 30
        self._year -= years
        self._month -= month_
        if self._month < 0:
            self._year -= 1
            self._month += 12
        self._day -= days
        if self._day < 30:
            if self._month == 1:
                self._month = 12
                self._year -= 1
            self._day += 30


class TDate1(TDate):
    def __init__(self, date):
        super().__init__()
        numbers = date.split('.')
        self._day = int(numbers[0])
        self._month = int(numbers[1])
        self._year = int(numbers[2])

    def __str__(self):
        return "{0}.{1}.{2}".format(self._day, self._month, self._year)


class TDate2(TDate):
    def __init__(self, date):
        super().__init__()
        numbers = date.split('-')
        self._day = int(numbers[1])
        self._month = int(numbers[0])
        self._year = int(numbers[2])

    def __str__(self):
        return "{0}-{1}-{2}".format(self._month, self._day, self._year)
